src/ln/channelmanager.rs

```diff
@@ -1988,6 +1994,39 @@ impl ChainListener for ChannelManager {
 		for failure in failed_channels.drain(..) {
 			self.finish_force_close_channel(failure);
 		}
+
+		{
+			let mut unsolved_htlcs = self.unsolved_htlcs.lock().unwrap();
+			for tx in txn_matched {
+				unsolved_htlcs.retain(|&payment_hash| {
+					let payment_hash160 = {
+						let mut ripemd = Ripemd160::new();
+						ripemd.input(&payment_hash);
+						let mut res = [0; 20];
+						ripemd.result(&mut res);
+						res
+					};
+					//TODO: size of redeemScript is test-only right now
+					if tx.input.len() > 0 && tx.input[0].witness.len() == 5 && tx.input[0].witness[4].len() == 138 && payment_hash160 == tx.input[0].witness[4][69..89] {
```

Comment:
Maybe somehow return them from ChannelMonitor instead?

Reply:
Well, I didn't change it because, here the match is done against claimable_htlcs from channel_manager, just having a wrapper to extract preimage in channel_monitor, I don't think I have access to funding_txo oupoint here to ask to the right channel_monitor, it's really _inter_-channels stuff for me.

Reply:
Sure, but a ChannelMonitor should know how to monitor for things that happened to its channel on chain, including an HTLC claim that we may not have any response to, but saw. Even the current implementation relies on the since-closed ChannelMonitor behaving correctly to register transactions for monitoring.

Reply:
Yeah, I obviously lay on ChannelMonitor for detection of revocations so will refactor with is_fulfilled, is_timeout, is_revoked_tx calls to ChannelMonitor, which will handle state detection ? Or do you think it's better to have states registered in a hashmap by ChannelMonitor at blocks parsing and just ChannelManager peeks them up with hash as a key ?

Reply:
I'm not quite sure what you're asking here, but clearly ChannelMonitor knows the relevant information to figure out the preimages and then pass them back up to ChannelManager.

Reply:
Sorry, more something like in the last commit I just added about the fail backward in case of revocation.